File: Python/ArrayAndString/pascal_triangle.py
# -*- coding: utf-8 -*-
# =============================================================================
# Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.
# Example:
# 
# Input: 5
# Output:
# [
#      [1],
#     [1,1],
#    [1,2,1],
#   [1,3,3,1],
#  [1,4,6,4,1]
# ]
# =============================================================================

def generate(numRows):
    """
    :type numRows:int
    :rtype List[List[int]]
    """
 
     if numRows == 0:
         return []
     result = []
     for i in range(numRows):
#         if i == 0:
#             result.append([1]) ## append 1
#         elif i == 1:
#             result.append([1, 1]) ## append 1, AND append 1 at the end
#         else:
             level = []
             level.append(1) # first 1  ## append 1
             for j in range(1, i):
                 level.append(result[i - 1][j -1] + result[i - 1][j])
#              level.append(1) # last 1  ## append 1 at tge end
             if i > 0:
                 level.append(1) # add last 1
             result.append(level)
     return result

 

if __name__ == '__main__':
    result = generate(5)
    for i in result:
        for j in result[i]:
            print(j)
        

File: Java/recursion/PascalTriangle.java
package recursion;

import java.util.ArrayList;
import java.util.List;

/*Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.


In Pascal's triangle, each number is the sum of the two numbers directly above it.

Example:

Input: 5
Output:
[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]
*/

public class PascalTriangle {

	public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> triangle = new ArrayList<List<Integer>>();
        if (numRows == 0)
        {
            return triangle;
        }
        
        for (int i = 0; i < numRows; i++)
        {
            if (i == 0 || i == 1)
            {
            	List<Integer> rows = new ArrayList<Integer>();
                for (int j = 0; j <= i; j++){
                    rows.add(1);
                }
                triangle.add(rows);
                continue;
            }
            
            List<Integer> rows = new ArrayList<Integer>();
            rows.add(1);//first item
            List<Integer> lastRow = triangle.get(i-1);//last layer
            for (int item = 1; item < i; item++){
                int middleValue = lastRow.get(item - 1) + lastRow.get(item);
                rows.add(middleValue);
            }
            rows.add(1);//last item
            
            triangle.add(rows);
        }
        return triangle;
    }
	
	// TLE!!!
	public List<List<Integer>> iterativeSol(int numRows) {
		List<List<Integer>> triangle = new ArrayList<List<Integer>>();
		if (numRows == 0) return triangle;
		for (int i = 1; i <= numRows; i++) {
			List<Integer> row = new ArrayList<>();
			for (int j = 1; j <= numRows; j++) {
				// get the each cell value using iterative solution
				row.add(helper(i, j));
			}
			triangle.add(row);
		}
		return triangle;
	}
	
	public int helper(int rowNum, int colNum) {
		if (rowNum == 1) return 1;
		if (colNum == 1 || colNum == rowNum) return 1;
		return helper (rowNum - 1, colNum - 1) + helper(rowNum - 1, colNum);
	}
}
